enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,971 @@
+is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
+To quote using single-quotes, we will do the following:
+Replace ' with '\''.
+Wrap the entire string in single quotes.
+eg: "abc"\\ 'def'  shall it be converted to '"abc"\\ '\''def'\'''
+
+
+### Risks and Mitigations
+
+We identified the following headers where some of them either can be allowed to set/delete whereas some headers need not to be allowed to set/delete
+due to the reasons mentioned in the following:
+
+#### Proxy header
+  As per haproxy security vulnerability, if the client sets `proxy` header, then the cgi app takes the value and injects it in the env var `HTTP_PROXY`,
+  so basically client is setting the proxy which is a security vulnerability so that is why we deny setting/deleting proxy header with this EP.
+  Please note that proxy headers are deleted by haproxy template by default.
+  `proxy` isn't necessarily a security risk (route owners could inject configuration into their own applications) but is non-standard and potentially
+  confusing, so it should be blocked.
+
+#### Strict-Transport-Security
+  strict-transport-security is a security risk (if we allowed it, then route owners could impact each other's routes), so it must be blocked so we have put this header in the deny list.
+  Please refer the question `Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this.
+  Would it not be better to provide a concrete API and deprecate the annotation API?` in the `Open Questions` section.
+
+#### X-Forwarded-For
+  `x-forwarded-for` and related headers are potentially confusing (as modifying these headers using this new API could affect configuration specified using
+   another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
+  but have actual use-cases (which aren't covered by the existing option), so they should be allowed, with notes in the API and product documentation,
+  and possibly a warning on admission.
+
+#### X-SSL
+ `x-ssl` and related headers similarly have an actual use-case, and allowing the cluster admin or route owner to modify these presents no security risk.
+  Eg: haproxy takes information from tls handshake or tls session or the certificate that client provided. These headers are used for mtls and client certificates.
+  These headers tell you did the client provide a certificate or was the certificate valid or what was the content or cipher or distinguished name or expiration date
+  on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent
+  to the backend application.
```

Comment:
This is not clear, needs some re-wording and a grammar check.  If I could figure it out, I'd make a suggestion, but I can't.

Reply:
I am not sure how to reword it more. If you can explain what is needed then that will be great.

Reply:
just made some grammar changes